src/cephadm/design_choices.rst

```diff
@@ -0,0 +1,165 @@
+Cephadm Design Docs
+===================
+
+
+**Housekeeping note:**
+
+If you want to see things added to this document, please add it via Github's `suggestion[Suggest Changes]` feature.
+This allows me to just hit the button in order for the comment to end up in the document.
+
+
+Intro
+-----
+
+Current situation:
+
+Cephadm manages all registered hosts. This means that it needs to periodically scrape data from each
+host to identify changes on the host like:
+
+* disk added/removed
+* daemon added/removed
+* host network/firewall etc has changed
+
+
+Currently, cephadm scrapes each host (up to 10 in parallel[0]) every 6 minutes[1], unless a refresh is forced manually.
+
+Refreshes for disks (ceph-volume), daemons (podman/docker), etc, happen in sequence.
+
+While the refresh model isn't ideal in the first place, the execution time vastly increases with the cluster size > 10
+
+This certainly isn't acceptable since ceph itself scales massively and so has the management part.
+
+There are a couple of solutions to this:
+
+
+1. Pre fetching/caching data on the host
+----------------------------------------
+
+This would allow bin/cephadm to autonomously scrape data that can later be retrieved by mgr/cephadm.
+
+i.e. `cephadm ls` would access an underlying (transparent) cache.
+
+This, btw assumes a **pull** model.
+
+There is at least one possible solutions on *how* to do this.
+
+1.1 daemonize bin/cephadm
+--------------------------
+
+A daemon is able to execute scraping tasks more frequently (as opposed to only when asked by mgr/cephadm)
+
+
+Downsides:
+
+We're still `scp'ing` bin/cephadm from the ceph-mgr to the host which is then executed.
+A daemon on the other hand needs to run continuously on the target system. If ceph-mgr (and the containing bin/cephadm
+script) is updated, the daemon needs to be informed of this event and get updated/restarted as well.
+
+
+1.1.1 HTTP:
+___________
+
+There is a implementation by Paul Cuzner [2] that uses HTTP to serve an endpoint to the hosts metadata.
+
+Upside:
+
+* We could utilize a proper REST API to CRUD all of the daemons/disks etc.
```

Comment:
The PR specifically introduces a command called "exporter" to avoid misunderstandings relating to CRUD functionality. In addition, the problem statement defined in the "Intro" is *not* talking about CRUD functionality - so it's weird that CRUD is referred to in many areas of this doc

Is the goal of this doc to define the architecture to support an agent based topology or to solve the metadata issue as defined in the Into? @jschmid1

Reply:
Imho the metadata issue naturally leads to considering future extensions of the approaches. This is also the goal of a design doc. It literally says `We could utilize...`

Reply:
> Is the goal of this doc to define the architecture to support an agent based topology or to solve the metadata issue as defined in the Into? @jschmid1

One leads to the other. If we're introducing a HTTP endpoint to retrieve metadata, the likelihood of using this HTTP endpoint for other operations is higher. Hence we need to evaluate the pros & cons _beforehand_.

Reply:
@jan-f Yes it does say "we could utilize"...but that's not in the problem statement at all. It was a statement from Joshua - I'm basically trying to help us focus on the problem statement. At the end of the day we could do anything!

Are you guys saying that the only acceptable way to improve the data quality to the orch and UI is to implement an agent architecture?

Reply:
I was saying that _if_ we go the HTTP route, one of the upsides of going this path is that we _may_, _eventually_, also use it to implement cephadm's other capabilities on top of it.

We've been talking for month about re-architecturing cephadm  to make it more testable, maintainable etc. This is just something that we need to take into account when making a change of this size.